labop/execution_engine.py

```diff
@@ -209,31 +219,40 @@ def run(
 
         # Iteratively execute all unblocked activities until no more tokens can progress
         while ready:
-            non_call_nodes = [
-                node
-                for node in ready
-                if not isinstance(node, uml.CallBehaviorAction)
-            ]
-            if non_call_nodes:
-                ready = self.step(non_call_nodes)
-            else:
-                ready = self.step(ready)
+            ready = self.step(ready)
         return ready
 
     def step(
         self,
         ready: List[uml.ActivityNode],
         node_outputs: Dict[uml.ActivityNode, Callable] = {},
     ):
-        for node in ready:
+        non_call_nodes = [
+            node
+            for node in ready
+            if not isinstance(node, uml.CallBehaviorAction)
+        ]
+        new_tokens = []
+        # prefer executing non_call_nodes first
+        for node in non_call_nodes + [
+            n for n in ready if n not in non_call_nodes
+        ]:
```

Comment:
Took me a minute to figure out what the list comprehensions were doing here.  Consider this one-liner to make the operation more explicit and readable:
```
# prefer executing non_call_nodes firs
ready.sort(key=lambda node: isinstance(node, uml.CallBehaviorAction))
```